Add Management Command for Batch Processing Items

rajadain opened the pull request:
## Overview

Adds a management command for batch processing of facility list items.

Connects #161 

## Demo

![image](https://user-images.githubusercontent.com/1430060/52298450-f8cfc800-2950-11e9-96e0-b840d0dba9f2.png)

## Notes

What should be done in the case of partial failures? For example, if geocoding fails for a few results, matching cannot be run at all. But neither can geocoding be re-run, leaving the user in a trapped state:

![image](https://user-images.githubusercontent.com/1430060/52298857-fd48b080-2951-11e9-9cdf-d7aaf158994a.png)

Should there be a flag to ignore error cases? Or to abort if there is a single error?

## Testing Instructions

* Check out this branch
* Run `manage batch_process --help`
* Run `manage batch_process --action [parse|geocode|match] --list-id [id]`
* Ensure it works correctly and handles failure cases

jwalgran commented:
Looking at this now

jwalgran commented:
>What should be done in the case of partial failures?

Good question. When we actually run these jobs using AWS Batch the plan is to have each row be processed as a separate Array Job instance so that they are all run in parallel. When a Job fails with an application-level error, like a geocode failure, we can choose to exit with a zero status code or a non-zero status code. I am not completely sure which is going to be the best course of action. If exiting non-zero causes the entire job or chain or of dependent batch jobs to be aborted, that is probably not what we want. Based on my reading of the documentation that is what will happen, so my current thinking is that we move the item with a failing geocode to the `ERROR` state and exit zero as we are doing. When the match Job for that item starts up it is able to fail fast on the status check and exits zero as well. The full Array Job will then be "successful" at the system level, and we can report on the finer-grained failures in the application.

There are different ways in which the individual tasks can fail. How we deal with retrying them is an open question that I think we will want to get feedback from our advisors on after we have the system up and running. It can be as simple as giving them the ability to move the rolling the `status` of an item back and queuing up a new, one-off chain of Jobs for that item.

rajadain commented:
Rebased atop `develop` to resolve merge conflicts.

That matches the current behavior then, where the exit code is always 0 regardless of the process succeeding or not:

<img width="517" alt="image" src="https://user-images.githubusercontent.com/1430060/52306929-38a1aa00-2967-11e9-988e-5387d716bb80.png">

(I did it from within Vagrant this time to be closer to the command whose exit code we're inspecting)

rajadain commented:
Thanks for catching those! Makes sense. Have a look at 29af85f and see if it covers it.

rajadain commented on src/django/api/management/commands/batch_process.py:
```diff
@@ -0,0 +1,104 @@
+                        result['failure'] += 1
+                    else:
+                        result['success'] += 1
+                except DatabaseError:
```
I'm catching a `DatabaseError` here because that's what `transaction.atomic()` will throw. `process` has error handlers inside it, and will catch any local errors.

rajadain commented on src/django/api/management/commands/batch_process.py:
```diff
@@ -0,0 +1,104 @@
+                            match.save()
+                    else:
+                        process(item)
+                        item.save()
```
Should this single save also be wrapped in `transaction.atomic()`?

rajadain commented:
Thanks for all the detailed and thoughtful feedback! I really appreciate it!

I added in your two commits to this branch.

I also added a final fixup b85c439 which does the following:

* Crashes immediately on incorrect `action` specification, rather than going through a `try` `raise` `except` dance
* Moves `transaction.atomic()` further up to encapsulate all the `process` and `.save()` calls
* Adds an `except ValueError` block within the loop to log and count those as failures, in a way that doesn't stop the larger process
* Updates the `except DatabaseError` block to log the error and crash

rajadain commented on src/django/api/management/commands/batch_process.py:
```diff
@@ -0,0 +1,110 @@
+                            process(item)
+                            item.save()
+
+                        if item.status == FacilityListItem.ERROR:
```
This probably doesn't need to be within `transaction.atomic()`

jwalgran replied:
Correct. We are done with the database at this point, so we could exit the `with` block here. Not a big deal since we aren't doing anything expensive before we start the next iteration.

rajadain commented on src/django/api/management/commands/batch_process.py:
```diff
@@ -0,0 +1,110 @@
+                            action, result['failure'])))
+
+        except FacilityList.DoesNotExist:
+            self.stderr.write('Validation Error: '
```
Should we also `sys.exit(1)` here?

jwalgran replied:
I think so. It makes sense to fail the Job the AWS Batch level if one of the arguments was wrong.

rajadain replied:
Should I do the same for `CommandError` beneath this?

rajadain commented:
The new commits seem to have made some tests fail though:

```
======================================================================
ERROR: test_matches_existing_facility (api.tests.FacilityListItemMatchingTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/django/db/backends/utils.py", line 83, in _execute
    return self.cursor.execute(sql)
psycopg2.IntegrityError: insert or update on table "api_facilitymatch" violates foreign key constraint "api_facilitymatch_facility_id_7280e334_fk_api_facility_id"
DETAIL:  Key (facility_id)=() is not present in table "api_facility".
```

Should there be a migration to go along with the new GUID stuff?

rajadain commented:
Alright, removed a bunch of unnecessary nesting in `try` and `sys.exit(1)` which canceled out nicely in 30fb6cf. Going to squash changes and merge when green. Thanks for all your help with this!
